### PR TITLE
Implement InstanceMetadataByProviderID

### PR DESCRIPTION
**Pull request by gongguan**
Ref [#90652](https://github.com/kubernetes/kubernetes/issues/90652)

**Special notes for your reviewer**:
Implement `InstanceMetadataByProviderID`

`InstanceMetadataByProviderID` used to get instance providerID, type and nodeAddressed in one function to reduce the number of calls to cloud API.
**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by gongguan** [comment]
/test pull-kubernetes-integration

azure tests failure are irrelevant to this pr.
`ERROR: azure-cli-core 2.2.0 has requirement knack~=0.6.2, but you'll have knack 0.7.1 which is incompatible.`

**Comment by gongguan** [comment]
/assign @wojtek-t 
I think it's ready for review.

**Comment by wojtek-t** [comment]
@gongguan -  thanks a lot for opening this PR; I will try to look into it in the next couple days; in the meantime, there are some azure-releated test failures, so would be great if you could take a look

**Comment by gongguan** [comment]
> there are some azure-releated test failures, so would be great if you could take a look

I think it caused by https://github.com/kubernetes/test-infra/commit/6f74cd9e7d558d00fd6d6d9889e3f4233266b5d4
Since aksengine-orchestratorRelease>=1.17, aksDeploymentMethod use customK8sComponents, obviously there is something wrong in [aks-engine 1.17](https://github.com/Azure/aks-engine/blob/master/docs/topics/kubernetes-developers.md#kubernetes-117)

In normal test https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90800/pull-kubernetes-e2e-azure-file-windows/1258016499126243328
kubetest run with `--aksengine-orchestratorRelease=1.15` 

FYI @chewong @feiskyer

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 1]
```diff
@@ -187,6 +187,8 @@ type Instances interface {
 	InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error)
 	// InstanceShutdownByProviderID returns true if the instance is shutdown in cloudprovider
 	InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error)
+	// InstanceMetadataByProviderID returns the instance's metadata which contains its type and nodeAddresses.
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:190]: nit: I would remove "which contains..." part. We may add something to it later and this will become outdated.

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 2]
```diff
@@ -187,6 +187,8 @@ type Instances interface {
...
+	InstanceMetadataByProviderID(ctx context.Context, providerID, nodeName string) (*InstanceMetadata, error)
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:191]: I suggest removing "nodeName" from the parameters (and leave only providerID).

In case providerID is empty (or cloud provider won't return anything for it), we can return error.

[Having just providerID would make it consistent with other methods too.]

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 3]
```diff
@@ -1646,6 +1646,11 @@ func (c *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
 	return false, nil
 }
 
+// InstanceMetadataByProviderID returns metadata of the specified instance.
+func (c *Cloud) InstanceMetadataByProviderID(ctx context.Context, providerID, nodeName string) (*cloudprovider.InstanceMetadata, error) {
+	return nil, nil
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Return "unimplemented error" for now.

Same in other places.

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 2 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go:191]: If so, do we need get instanceMetadata by nodeName?

Comment (wojtek-t) [reply staging/src/k8s.io/cloud-provider/cloud.go:191]: Maybe - but let's maybe start with just byProviderId - I'm still not convinced it will be really needed.

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 1 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go]: Addressed.

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 2 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go]: Done, looks easier.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 3 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Done.

File: pkg/volume/cinder/attacher_test.go [hunk 4]
```diff
@@ -734,6 +734,10 @@ func (instances *instances) InstanceShutdownByProviderID(ctx context.Context, pr
 	return false, errors.New("unimplemented")
 }
 
+func (instances *instances) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
+	return &cloudprovider.InstanceMetadata{}, errors.New("unimplemented")
```
Comment (wojtek-t) [inline pkg/volume/cinder/attacher_test.go:738]: return nil and error

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 5]
```diff
@@ -184,6 +184,8 @@ type Instances interface {
 	InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error)
 	// InstanceShutdownByProviderID returns true if the instance is shutdown in cloudprovider
 	InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error)
+	// InstanceMetadataByProviderID returns the instance's metadata.
+	InstanceMetadataByProviderID(ctx context.Context, providerID string) (*InstanceMetadata, error)
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:188]: Once we have this, we should migrate all calls to NodeAddressesByProviderId and InstanceTypeByProviderId to this call.
So can you please add a comment:
// Deprecated: Remove once all calls are migrated to InstanceMetadataByProviderID
to both these methods.

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 6]
```diff
@@ -250,3 +252,13 @@ type Zones interface {
 type PVLabeler interface {
 	GetLabelsForVolume(ctx context.Context, pv *v1.PersistentVolume) (map[string]string, error)
 }
+
+// InstanceMetadata is a collection of instance's type and nodeAddressed.
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:256]: Let's not mention specific fields explicitly here - adding a new field will make this comment outdated.
Maybe "InstanceMetadata contains metadata about the specific instance".

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go [hunk 7]
```diff
@@ -163,7 +163,40 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
...
-	return nil, fmt.Errorf("unimplemented")
+	md := &cloudprovider.InstanceMetadata{}
+	var err error
+
+	if providerID == "" {
+		return nil, fmt.Errorf("no providerID")
+	}
+
+	md.ProviderID = providerID
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:174]: nit: let's move it to the top of the function (consistency with the rest of this file)

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go [hunk 8]
```diff
@@ -163,7 +163,40 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
 
 // InstanceMetadataByProviderID returns metadata of the specified instance.
 func (g *Cloud) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
-	return nil, fmt.Errorf("unimplemented")
+	md := &cloudprovider.InstanceMetadata{}
+	var err error
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:167]: Seems unneeded

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go [hunk 9]
```diff
@@ -163,7 +163,40 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
...
+
+	if providerID == "" {
+		return nil, fmt.Errorf("no providerID")
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:170]: nit: "couldn't compute InstanceMetadata for empty providerID"

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go [hunk 10]
```diff
@@ -163,7 +163,40 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
...
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := g.c.Instances().Get(timeoutCtx, meta.ZonalKey(canonicalizeInstanceName(name), zone))
+	if err != nil {
+		return nil, fmt.Errorf("error while querying for providerID %q: %v", providerID, err)
+	}
+
+	if len(instance.NetworkInterfaces) < 1 {
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:187]: Can you please factor this out to a separate function (nodeAddressesFromGCEInstance) and share it between this and the NodeAddressesByProviderID functions?

Comment (wojtek-t) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:187]: It seems the same comment holds for some other cloud providers (vsphere is what I looked at).

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 5 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go:188]: I plan to remove these in follow-up work(when node-controller uses InstanceMetadataByProviderID). It's ok to add this comment.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go [hunk 11]
```diff
@@ -163,7 +153,43 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
 
 // InstanceMetadataByProviderID returns metadata of the specified instance.
 func (g *Cloud) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
-	return nil, fmt.Errorf("unimplemented")
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+
+	md := &cloudprovider.InstanceMetadata{}
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go:159]: nit: instead of defining it here, I think the cleaner flow would be:
```
...
nodeAddresses, err := nodeAddressesFromInstance(instance); err != nil {
  return nil, err
}
return &cloudprovider.InstanceMetadata{
  ProviderID: providerID,
  Type: lastComponent(instanceMachineType),
  NodeAddresses: nodeAddresses,
}, nil
```

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go [hunk 12]
```diff
@@ -772,7 +753,106 @@ func (vs *VSphere) InstanceShutdownByProviderID(ctx context.Context, providerID
...
+	if vclib.IsManagedObjectNotFoundError(err) {
+		if err = vs.nodeManager.RediscoverNode(k8stypes.NodeName(nodeName)); err != nil {
+			return nil, err
+		}
+		return vs.instanceMetadataByNodeName(ctx, nodeName)
+	}
+	return md, err
+}
+
+func (vs *VSphere) instanceMetadataByNodeName(ctx context.Context, nodeName string) (*cloudprovider.InstanceMetadata, error) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:775]: I'm not familiar with this code, but it seems you can probably share a bit more than the nodeAddressesFromVM part

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go [hunk 13]
```diff
@@ -583,68 +583,43 @@ func testHasNodeAddress(t *testing.T, addrs []v1.NodeAddress, addressType v1.Nod
...
 
-func TestNodeAddresses(t *testing.T) {
-	// Note instance0 and instance1 have the same name
-	// (we test that this produces an error)
-	var instance0 ec2.Instance
-	var instance1 ec2.Instance
-	var instance2 ec2.Instance
-
-	// ClusterID needs to be set
+func makeInstance(num int, privateIP, publicIP, privateDNSName, publicDNSName string) ec2.Instance {
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go:586]: nit: might be useful to send this refactoring as a separate PR (it's non-controversial and would make this commit pretty simple)

File: pkg/volume/cinder/attacher_test.go (hunk 4 above)
Comment (gongguan) [reply pkg/volume/cinder/attacher_test.go]: Done.

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 5 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go:191]: Done.

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 6 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go]: Done, it's good.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go (hunk 8 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go]: Dropped

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go (hunk 9 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go]: done in all implementation.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go (hunk 11 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go]: addressed in all implementations, except for azure(makes multiple return cleaner)

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 12 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:803]: You're right, all logic of `NodeAddresses` can be reused, cleaner now I think.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go (hunk 13 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go:586]: Done, see [#91082](https://github.com/kubernetes/kubernetes/pull/91082)

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 14]
```diff
@@ -1646,6 +1646,32 @@ func (c *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
...
+	instanceID, err := KubernetesInstanceID(providerID).MapToAWSInstanceID()
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := describeInstance(c.ec2, instanceID)
+	if err != nil {
+		return nil, err
+	}
+	if *instance.State.Name == ec2.InstanceStateNameTerminated {
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1660]: It seems this check isn't there neither in NodeAddressesByProviderID nor in InstanceTypeByProviderID

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go [hunk 15]
```diff
@@ -770,6 +778,51 @@ func (vs *VSphere) InstanceShutdownByProviderID(ctx context.Context, providerID
 	return !isActive, nil
 }
 
+// InstanceMetadataByProviderID returns metadata of the specified instance.
+func (vs *VSphere) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
+	if providerID == "" {
+		return nil, fmt.Errorf("couldn't compute InstanceMetadata for empty providerID")
+	}
+
+	nodeName, err := vs.GetNodeNameFromProviderID(providerID)
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:787]: It seems NodeAddressesByProviderID is converting it differently:
https://github.com/kubernetes/kubernetes/blob/0f2cccc98c3bf2a512ec55cbeba89dbfe7c768cf/staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go#L707

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 12 above)
Comment (wojtek-t) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:802]: Thanks!

**Review by wojtek-t — COMMENTED** [review]
Modulo the minor comments that I just added everything by azure stuff looks fine.
We need to find someone to review Azure stuff though (other providers are pretty straightforward).

@feiskyer - maybe

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 14 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1660]: It's a logic of [InstanceExistsByProviderID](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L1605).
I had thought it's useful to judge whether an instance exists.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 15 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:787]: It's a logic from InstanceExistsByProviderID too. It will be cleaner to use providerID as nodeName directly if it won't cause any issue(I'm not sure now).

**Comment by wojtek-t** [comment]
/assign @feiskyer 
Can you please help with reviewing the Azure part?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 14 above)
Comment (wojtek-t) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1660]: Yeah - I've seen that. But it wasn't used in NodeAddressesByProviderID.

So to make this a no-op change (which is really a goal now), let's remove it for now.
Same below.

You can add a TODO to validate if we don't want to add it.

Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: all done.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go [hunk 16]
```diff
@@ -770,6 +778,49 @@ func (vs *VSphere) InstanceShutdownByProviderID(ctx context.Context, providerID
...
 
+// InstanceMetadataByProviderID returns metadata of the specified instance.
+func (vs *VSphere) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
+	if providerID == "" {
+		return nil, fmt.Errorf("couldn't compute InstanceMetadata for empty providerID")
+	}
+
+	// TODO not get nodeName by GetNodeNameFromProviderID here. If it not behave expected,
+	// get nodeName by vm.GetNodeNameFromProviderID.
+	nodeName := providerID
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:789]: Let's use: `convertToK8sType` (as it is used in NodeAddressesByProviderID)

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go [hunk 17]
```diff
@@ -770,6 +778,49 @@ func (vs *VSphere) InstanceShutdownByProviderID(ctx context.Context, providerID
...
+	md, err := vs.instanceMetadataByNodeName(ctx, nodeName)
+	if vclib.IsManagedObjectNotFoundError(err) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:791]: It wasn't there in NodeAddresses nor NodeAddressesByProviderID.

let's remove (at least for now) - to make that a no-op change.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 16 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: done.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 17 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: done.

**Comment by gongguan** [comment]
kind ping @feiskyer @andyzhangx @khenidak for azure area review.

TL;DR;
InstanceMetadataByProviderID get instance type, existence and nodeAddresses which reduce the number of call cloud api. (compare with InstanceExistsByProviderID, InstanceTypeByProviderID and NodeAddressesByProviderID step by step in cloud/node-controller).

**Review by feiskyer — APPROVED** [review]
looks good for the Azure changes.

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Comment by andrewsykim** [comment]
I'm generally not opposed to the changes made here, but any changes we make to the cloud provider interface should be discussed in a SIG (cloud provider) meeting first. There are many downstream effects to every addition of the cloud provider APIs and we want to give providers (both in-tree and out-of-tree) a chance to provide feedback. cc @cheftako

**Comment by gongguan** [comment]
@andrewsykim sorry for not reporting this. We're not intend to break downstream projects.

**Comment by andrewsykim** [comment]
np! I just want to make sure providers are aware of the implications of the changes here. I will raise it in the next SIG meeting :)

**Comment by gongguan** [comment]
Thanks Andrew, current implications are only marked as deprecated.

**Comment by andrewsykim** [comment]
Based on changes in https://github.com/kubernetes/kubernetes/pull/91319, removing the callers to deprecated methods is no different than removing the methods entirely.
